### PR TITLE
Install longhorn-nfs before all rwx cases

### PR DESCRIPTION
**Pull request by chanow816**
Our RWX caes were flaky and often failed at early test step related to statefulset.
We think this may due to nfs-common not installed well on all nodes
This PR create an fixture to install longhorn-nfs before each RWX cases to see if the flaky will improved.

Task for tracking rwx flaky cases: https://github.com/longhorn/longhorn/issues/2748

cc @khushboo-rancher

### CODE REVIEW

File: manager/integration/tests/common.py [hunk 1]
```diff
@@ -4555,3 +4561,42 @@ def restore_backup_and_get_data_checksum(client, core_api, backup, pod,
...
+    for i in range(RETRY_COUNTS):
+        nfs_pods = {}
+        pod_list = api.list_namespaced_pod("default")
+
+        for pod in pod_list.items:
+            pod_name = pod.metadata.name
+            if "longhorn-nfs" in pod_name:
+                nfs_pods[pod_name] = \
+                    api.read_namespaced_pod(name=pod_name,
+                                            namespace='default').status.phase
```
Comment (jenting) [inline manager/integration/tests/common.py:4585]: Probably you could consider using the label selector to filter longhorn-nfs Pods. For example
```shell
kubectl get pod -l app=longhorn-nfs-installation
```

**Review by khushboo-rancher — CHANGES_REQUESTED** [review]
Can you please use fixture with module scope so that at the beginning of the tests, only once nfs-common can be installed and removed after all the test cases have ran.
`@pytest.fixture(scope="module", autouse="True")`

**Comment by innobead** [comment]
Could we have this kind of setup once instead of running in each nfs test case via fixture? It seems repeated applying install manifest even the package is already installed after the first applying. WDYT?

Do we have global setup before running test suites/cases? Also what's the cause the nfs package sometimes can't be installed via terraform?

File: manager/integration/tests/test_rwx.py [hunk 2]
```diff
@@ -30,7 +30,7 @@ def write_data_into_pod(pod_name_and_data_path):
                                  DATA_SIZE_IN_MB_3)
 
 
-def test_rwx_with_statefulset_multi_pods(core_api, statefulset):  # NOQA
+def test_rwx_with_statefulset_multi_pods(nfs, core_api, statefulset):  # NOQA
```
Comment (khushboo-rancher) [inline manager/integration/tests/test_rwx.py:33]: If you're using the fixture, no need to call it for single test. Module here is all the tests in the file test_rwx, so the fixture with scope module will run automatically before starting of any test in this file.

File: manager/integration/tests/test_rwx.py [hunk 3]
```diff
@@ -96,7 +96,7 @@ def test_rwx_with_statefulset_multi_pods(core_api, statefulset):  # NOQA
         assert pod_data == md5sum_pod[i]
 
 
-def test_rwx_multi_statefulset_with_same_pvc(core_api, pvc, statefulset, pod):  # NOQA
+def test_rwx_multi_statefulset_with_same_pvc(nfs, core_api, pvc, statefulset, pod):  # NOQA
```
Comment (khushboo-rancher) [inline manager/integration/tests/test_rwx.py:99]: Same as above comment.

File: manager/integration/tests/test_rwx.py [hunk 4]
```diff
@@ -159,7 +159,7 @@ def test_rwx_multi_statefulset_with_same_pvc(core_api, pvc, statefulset, pod):
         core_api, command2, share_manager_name, LONGHORN_NAMESPACE)
 
 
-def test_rwx_parallel_writing(core_api, statefulset, pod):  # NOQA
+def test_rwx_parallel_writing(nfs, core_api, statefulset, pod):  # NOQA
```
Comment (khushboo-rancher) [inline manager/integration/tests/test_rwx.py:162]: Same as above comment.

File: manager/integration/tests/test_rwx.py [hunk 5]
```diff
@@ -224,7 +224,7 @@ def test_rwx_parallel_writing(core_api, statefulset, pod):  # NOQA
     assert md5sum2 == share_manager_data2
 
 
-def test_rwx_statefulset_scale_down_up(core_api, statefulset):  # NOQA
+def test_rwx_statefulset_scale_down_up(nfs, core_api, statefulset):  # NOQA
```
Comment (khushboo-rancher) [inline manager/integration/tests/test_rwx.py:227]: Same as above and applicable for all the below tests as well.

**Review by khushboo-rancher — CHANGES_REQUESTED** [review]

File: manager/integration/tests/test_rwx.py [hunk 6]
```diff
@@ -15,7 +15,7 @@
 from common import wait_delete_pod, wait_for_pod_remount
 from common import get_core_api_client, write_pod_volume_random_data
 from common import create_pvc_spec, make_deployment_with_pvc  # NOQA
-from common import core_api, statefulset, pvc, pod, client  # NOQA
+from common import core_api, statefulset, pvc, pod, client, nfs  # NOQA
```
Comment (jenting) [inline manager/integration/tests/test_rwx.py:18]: This line change is no more needed.

File: manager/integration/tests/common.py [hunk 7]
```diff
@@ -4555,3 +4561,43 @@ def restore_backup_and_get_data_checksum(client, core_api, backup, pod,
...
+
+        if len(nfs_pods) == len(node_list.items):
+            if all(value == "Running" for value in nfs_pods.values()):
+                break
+
+        time.sleep(RETRY_INTERVAL)
+
+    if not \
+            all(value == "Running" for value in nfs_pods.values()) \
+            or len(nfs_pods) == 0:
```
Comment (jenting) [inline manager/integration/tests/common.py:4596]: Should it be?
```python
or len(nfs_pods) == len(node_list.items):
```

File: manager/integration/tests/test_rwx.py (hunk 2 above)
Comment (chanow816) [reply manager/integration/tests/test_rwx.py]: Hello @khushboo-rancher, I kept import fixture nfs and removed all nfs function call.

File: manager/integration/tests/common.py (hunk 7 above)
Comment (chanow816) [reply manager/integration/tests/common.py]: I modified to `or not len(nfs_pods) == len(node_list.items)`

**Comment by chanow816** [comment]
Hi @innobead ,

I modified code, now longhorn-nfs will just install one time before module test_rwx started.

When creating nodes, terraform have command to install nfs-common but I didn't found code to check installation status. `def nfs()` will check longhorn-nfs install successfuly on all nodes. thank you.

**Review by innobead — APPROVED** [review]
LGTM

**Review by jenting — APPROVED** [review]
LGTM

**Comment by innobead** [comment]
@chanow816 please backport to 1.2.x or even 1.1.3.

**Pull request merged**